actionview/lib/action_view/helpers/tag_helper.rb

```diff
@@ -72,7 +74,8 @@ def #{method_name}(content = nil, escape: true, **options, &block)
                   TEXT
                   tag_string("#{name}", content, options, escape: escape, &block)
                 else
-                  self_closing_tag_string("#{name}", options, escape, ">")
+                  tag_suffix = ActionView::Helpers::TagHelper.void_element_trailing_slash ? " />" : ">"
+                  self_closing_tag_string("#{name}", options, escape, tag_suffix)
```

Comment:
Are there negative performance impacts from introducing this conditional into each individual call?

Is it possible to extract the suffix String definition so that it's outside the `batch.push(<<~RUBY)` string, then interpolated _back into_ the method String?

Reply:
@skipkayhil You are correct that it changes current behavior.

But without this change, e.g. these elements won't have the closing slash even if `config.action_view.void_element_trailing_slash` is set to `true`:
![Void elements without closing slash](https://github.com/rails/rails/assets/864340/a5f9f638-325c-4e07-9d92-7e2c082cb94a)

This source is generated by following the getting started guide of Rails.

It is fine by me to revert the change but then we are not respecting the configuration option in all cases, which was part of the fixes implemented after the previous review.

It's just a question what we want here:

1. Keeping the current behavior of having mixed void HTML tags (what is requested here)
2. Respecting the new configuration option in all cases (what was requested in the review)

Reply:
Depending on how https://github.com/rails/rails/pull/51646/files#r1578010728 is resolved, could the `tag_suffix` definition be inlined into the interpolation?

Then, that value could be interpolated directly into the evaluated Ruby method:

```suggestion
                  
                  self_closing_tag_string("#{name}", options, escape, "#{
                    ActionView::Helpers::TagHelper.void_element_trailing_slash ? " />" : ">"
                  }")
```

Reply:
Furthermore, is there a reason that the `void_element_trailing_slash` attribute need to be defined on `ActionView::Helpers::TagHelper`? Could it be defined on the `ActionView::Helpers::TagHelper::TagBuilder` class directly?

If the `mattr_accessor :void_element_trailing_slash, default: false` line were moved to the `TagBuilder` class, that configuration could be accessed directly from the class method, instead of as a fully-qualified class variable:

```suggestion
                  self_closing_tag_string("#{name}", options, escape, "#{void_element_trailing_slash ? " />" : ">"}")
```

Reply:
> Depending on how https://github.com/rails/rails/pull/51646/files#r1578010728 is resolved, could the tag_suffix definition be inlined into the interpolation?

Yes this is of course possible. But I'll just stress that it has no implication on performance at least to my knowledge.

I can do this change if it is insisted but I will wait for the feedback regarding the question on how to proceed with this, as there is the question above which is the desired direction with this.

> Furthermore, is there a reason that the void_element_trailing_slash attribute need to be defined on ActionView::Helpers::TagHelper? Could it be defined on the ActionView::Helpers::TagHelper::TagBuilder class directly?

It is also needed at `TagHelper#tag` method.

I don't have a preference where the configuration option is defined, I just thought it would be natural at the lowest nesting level where it is needed.

Reply:
@skipkayhil @rafaelfranca Could you chime in which would be the preferred direction forward, so that we could finalize this.

So there are two alternatives, as pointed out above:

1. Not to change this method which would make the HTML inconsistent regarding the introduced configuration flag, i.e. when the ending slashes are preferred, the HTML would be inconsistent and not respect the configuration option in all places.
2. Following the current direction where the configuration option is respected everywhere, but implement the changes requested here

Which one both of you think is the desired approach?